### ISSUE TITLE
'Invalid Date' and 'Reply' not working

### ISSUE THREAD

**Issue by xEnigma** [Viima/jquery-comments#2]
Hello,

I'm storing the comments to a database with Laravel via a post request. 
Then I'm loading then again with a get request. However, first when I send the comment, a new item appears in the list but there is no text. When I refresh the page it appears. The comment is written correctly to the database.
When I'm loading the comments from the GET Request, everything works so far, but I can't get the date working (always displays "Invalid Date") and the reply buttons are not working anymore.
When I use the example without post and get requests as you have described in "2) Initialize the library", everything is working so far. But of course I want to save the comments to the database.
When I use the example without post and get requests as you have described in "2) Initialize the library", everything is working so far. But of course I want to save the comments to the database.

Here is some code:

        $('.comments_content').comments({
            profilePictureURL: 'https://app.viima.com/static/media/user_profiles/user-icon.png',
        		enableUpvoting: false,
            roundProfilePictures: true,
            enableReplying: true,

            fieldMappings: {
              id: 'id',
              parent: 'parent',
              created: 'created_at',
              modified: 'updated_at',
              content: 'content',
              fullname: 'author_id',
              profilePictureURL: 'profile_picture_url',
              createdByCurrentUser: 'created_by_current_user'
            },
            postComment: function(commentJSON, success, error) {
                $.ajax({
                    type: 'post',
                    url: "{{ action('CommentsController@store') }}",
                    data: {comment: commentJSON, cardid: clickedcardid},
                    success: function(comment) {
                        success(comment)
                    },
                    error: error
                });
            }
        });

            getComments: function(success, error) {
                $.ajax({
                    type: 'get',
                    url: '{{ action('CommentsController@index') }}',
                    success: function(commentsArray) {

                      // Only display the comments which are assigned to the clicked card
                      for(var i = commentsArray.length - 1; i >= 0; i--) {
                        if(commentsArray[i].card_id != clickedcardid) {
                          commentsArray.splice(i, 1);
                        }
                      }

                      for(var i=0; i<commentsArray.length; i++) {
                        commentsArray[i].profile_picture_url = 'https://app.viima.com/static/media/user_profiles/user-icon.png';
                      }

                      success(commentsArray)
                    },
                    error: error
                });
            }

So to sum up:
- I can save the comments to a database and read them out
- I do it via jQuery AJAX POST and GET Requests
- Problem 1: When I post a comment, a new item appears in the list, however without the content and author displayed. First displayed after refresh.
- Problem 2: When reading out the comments, it always says "Invalid Date"
- Problem 3: When reading out, the reply button is not working.

Thanks for your help!

**Comment by jessenieminen** [comment]
Hi and thanks for the report @xEnigma !

**Regarding Problem 1:**
I can't reproduce the issue locally, could you provide a codePen or similar where the issue would be replicated?

**Regarding problems 2 and 3: **
I would say that these are most likely the result of your database/API providing the information (probably for the timestamp) in an incorrect format, which results in those errors. Here's an example of a JSON payload with the formats that work: https://app.viima.com/api/customers/1/comments/

Could you compare your format to the link above?

**Comment by webbird** [comment]
I am having the date format problem, too. The string sent to the server side script looks like this:

id=c3&parent=&created=2015-07-27T13%3A40%3A22.767Z&modified=2015-07-27T13%3A40%3A22.767Z&content=asdfasdfasdf
&fullname=You&profile_picture_url=&created_by_current_user=true&upvote_count=0&user_has_upvoted=false

More readable example: 2015-07-27T13:40:22.767Z

An example entry I created directly in the database with date format 2015-07-27 works just fine.

The callback is exactly the one from the docs, just with different script URI:

                postComment: function(commentJSON, success, error) {
                    $.ajax({
                        type: 'post',
                        url: CAT_URL+'/modules/jquery_comments/ajax/ajax_post_comment.php',
                        data: commentJSON,
                        success: function(comment) {
                            success(comment)
                        },
                        error: error
                    });
                }

**Comment by webbird** [comment]
var time = new Date().toISOString();

**Comment by jjtykkyl** [comment]
I believe that the problem is that you are trying to pass an escaped timestamp to the library which tries to create a Date object out of it. For example the timestamp you used as an example (2015-07-27T13:40:22.767Z) is unescaped and can be used as such. However, the timestamp that you are passing to the server is escaped (2015-07-27T13%3A40%3A22.767Z) and javascript is not able to create a native date object out of an escaped string.

It's a good practice to fix this problem on your server so that the data provided is always unescaped. Thus the front-end can always be sure that the data is ready to be processed at any given time. If you'd like to test the solution first before making any changes to the backend, you can post process the comment object in success callback of postComment:

                        comment.created = unescape(comment.created)
                        comment.modified = unescape(comment.modified)

**Comment by webbird** [comment]
In my case I fixed this by setting the date in my AJAX script, just ignoring what I get via POST. ;)

**Comment by jjtykkyl** [comment]
Sure the jquery-comments creates the date string at 583
new Date().toISOString()
but I don't think that's the one escaping the string or is it?

**Comment by webbird** [comment]
It's the AJAX call that escapes it. (Which is correct.) The problem seems to be how the result is handeled by the success() method. Maybe I am doing it wrong, but my AJAX script does not return anything.

**Comment by jessenieminen** [comment]
Hi again @webbird ! 

What the AJAX call returns is determined in your backend, so please make sure your API posts standard success and error callbacks. Success callbacks typically have the newly created or modified JSON element as a parameter, which lets you do the necessary updates in the front end based on what your application logic is like.

**Comment by sthapa123** [comment]
@webbird How did you fix the issue?? I'm having the same issue.

**Comment by sthapa123** [comment]
The issue has been resolved. I used the following and it's working now.

`postComment: function (commentJSON, success, error) {
                $.ajax({
                    type: 'post',
                    url: '',
                    dataType: 'json',
                    contentType: "application/json; charset=utf-8",
                    data: JSON.stringify(commentJSON),
                    success: setTimeout(function () {
                        success(commentJSON);
                                           }, 500),
                    error: error
                });
            },`
